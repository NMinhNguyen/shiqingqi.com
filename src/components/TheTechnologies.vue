<template>
  <div class="the-technologies" :style="technologiesStyle">
    <span class="the-technologies__write">
      {{ $store.getters.getText("I_WRITE") }}
    </span>
    <div class="the-technologies__list">
      <span
        v-for="(tech, i) in technologies"
        :key="`the-technologies-${i}`"
        :style="{ '--the-technologies-index': i }"
      >
        {{ tech }}
      </span>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    technologies: {
      type: Array,
      /** @returns {any[]} */
      default: () => ["JavaScript", "CSS", "HTML"]
    },
    animateScroll: { type: Number, default: 0 }
  },
  computed: {
    /** @returns {{[property: string]: number}} */
    technologiesStyle() {
      return {
        "--the-technologies-animate-scroll": this.animateScroll,
        "--the-technologies-count": this.technologies.length - 1
      };
    }
  }
};
</script>

<style lang="scss">
.the-technologies {
  display: flex;
  padding: 30vh 2rem 2rem;
  height: 100%;
  font-size: 2rem;

  &__write {
    flex-grow: 1;
    text-align: right;
    padding-right: 0.5rem;
  }

  &__list {
    $animate: var(--the-technologies-animate-scroll);
    flex-grow: 1;
    display: flex;
    flex-direction: column;
    height: 100%;
    justify-content: space-between;
    font-family: $font-heading;
    position: relative;
    transform: translateY(calc(#{$animate} * (-70vh + 4rem + 5px)));

    span {
      $n: var(--the-technologies-count);
      $i: var(--the-technologies-index);
      --the-technologies-fade-count: calc((#{$n} + 1) / 3);
      $f: var(--the-technologies-fade-count);

      opacity: calc(-1 * (#{$n} / #{$f}) * (#{$animate} - #{$i} / #{$n}) + 1);
    }
  }
}
</style>
