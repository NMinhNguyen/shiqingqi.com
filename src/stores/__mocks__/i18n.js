export default {
  state: {
    locale: "en",
    translations: {
      en: {
        test: "test"
      }
    }
  },
  mutations: {
    setLocale: jest.fn(),
    setTranslations: jest.fn()
  },
  actions: {
    fetchTranslations: jest.fn()
  },
  getters: {
    getText: jest.fn(() => () => "")
  }
};
