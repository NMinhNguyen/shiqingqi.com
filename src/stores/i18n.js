import Vue from "vue";

export const state = {
  locale: "en",
  /** @type {{[locale:string]: {[key:string]: string}}} */
  translations: {}
};

export const mutations = {
  /**
   * @param {typeof state} state State object
   * @param {string} locale Locale code
   */
  setLocale(state, locale) {
    state.locale = locale;
  },

  /**
   * @param {typeof state} state State object.
   * @param {{lang: string, translations: object}} payload New translations.
   */
  setTranslations(state, { lang, translations }) {
    Vue.set(state.translations, lang, {
      ...state.translations[lang],
      ...translations
    });
  }
};

export const actions = {
  /**
   * @param {import("vuex").ActionContext<typeof state, {}>} context
   * @param {string} lang Language code.
   * @returns {Promise}
   */
  async fetchTranslations({ commit }, lang) {
    /** @type {{default: object}} */
    const module = await import(`@/constants/translations/${lang}.json`);
    commit("setTranslations", { lang, translations: module.default });
  }
};

export const getters = {
  /**
   * @param {typeof state} state
   * @returns {(key: string, params: {[key:string]: string}) => string}
   */
  getText: state => (key, params) => {
    const translations = state.translations[state.locale];

    // Return empty string if translation doesn't exist
    if (!translations || !translations[key]) return "";

    // Return translation if no params provided
    if (!params) return translations[key];

    // Replace the params
    let text = translations[key];
    for (let name in params) {
      text = text.replace(`{{${name}}}`, params[name]);
    }
    return text;
  }
};

export default {
  state,
  mutations,
  actions,
  getters
};
