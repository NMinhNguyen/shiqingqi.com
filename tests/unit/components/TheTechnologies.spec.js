// @ts-nocheck
import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import TheTechnologies from "@/components/TheTechnologies.vue";
import i18n from "@/stores/i18n";

jest.mock("@/stores/i18n");

let localVue = createLocalVue();
localVue.use(Vuex);

describe("TheTechnologies", () => {
  let store;

  beforeAll(() => {
    store = new Vuex.Store({
      modules: {
        i18n
      }
    });
  });

  it("renders list of technologies", () => {
    expect(
      shallowMount(TheTechnologies, {
        localVue,
        store,
        propsData: {
          technologies: ["testA", "testB"]
        }
      })
    ).toMatchSnapshot();
  });

  it("default list of technologies", () => {
    expect(
      shallowMount(TheTechnologies, {
        localVue,
        store
      })
    ).toMatchSnapshot();
  });
});
