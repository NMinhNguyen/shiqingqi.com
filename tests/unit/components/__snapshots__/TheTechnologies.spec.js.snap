// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheTechnologies default list of technologies 1`] = `
<div class="the-technologies"><span class="the-technologies__write">

  </span>
  <div class="the-technologies__list"><span>
      JavaScript
    </span><span>
      CSS
    </span><span>
      HTML
    </span></div>
</div>
`;

exports[`TheTechnologies renders list of technologies 1`] = `
<div class="the-technologies"><span class="the-technologies__write">

  </span>
  <div class="the-technologies__list"><span>
      testA
    </span><span>
      testB
    </span></div>
</div>
`;
